src/message_parser.h

```diff
@@ -0,0 +1,217 @@
+        int offset = 0;
+
+        if (direct) {
+            // message accumulation not started
+            recv_data.cur_addr = buf;
+        }
+
+        while (len > 0) {
+            int chunk_len = len;
+            if (unlikely(!direct)) {
+                // (1) copy data chunk to accumulation buffer
+                chunk_len = std::min(recv_data.cur_size, len);
+                AccumulationStrategy::accumulate_more(recv_data, buf + offset, chunk_len);
+            }
+            int nprocessed;
+            process_chunk(recv_data, chunk_len, nprocessed);
+            len -= nprocessed;
+            offset += nprocessed;
+
+            if (recv_data.cur_offset == 0) {
+                // message is ready to be processed
+                bool ok = callback.handle_message();
+                if (unlikely(!ok)) {
+                    return false;
+                }
+                buf += offset;
+                if (likely(direct)) {
+                    recv_data.cur_addr += offset;
+                } else {
+                    // switch back from accumulation buffer to direct input buffer
+                    recv_data.cur_addr = buf;
+                    direct = true;
+                }
+                offset = 0;
+            }
+        }
+
+        if (likely(recv_data.cur_offset == 0)) {
+            // reset read offset to the start of the buffer
+            AccumulationStrategy::reset_read_buffer(recv_data);
+        } else if (likely(direct)) {
+            // Pending message is in a direct buffer => copy it into accumulation buffer.
+            // If message is already in accumulation buffer then copy is already performed in (1).
+            AccumulationStrategy::start_accumulation(recv_data, buf, recv_data.cur_offset);
+        }
+
+        return true;
+    }
+
+    inline void process_chunk(SocketRecvData &recv_data, int nbytes, int &nprocessed)
+    {
+        nprocessed = 0;
+        /* 1: message header is not received yet */
+        if ((recv_data.cur_offset + nbytes) < MsgHeader::EFFECTIVE_SIZE) {
+            recv_data.cur_size -= nbytes;
+            recv_data.cur_offset += nbytes;
+            nprocessed = nbytes;
+
+            /* 4: set current buffer size to size of remained part of message header to
+             *    guarantee getting full message header on next iterations
+             */
+            if (recv_data.cur_size < MsgHeader::EFFECTIVE_SIZE) {
+                recv_data.cur_size = MsgHeader::EFFECTIVE_SIZE - recv_data.cur_offset;
+            }
+            return;
+        } else if (recv_data.cur_offset < MsgHeader::EFFECTIVE_SIZE) {
+            /* 2: message header is got, match message to cycle buffer */
+            m_pMsgReply->setBuf(recv_data.cur_addr);
+            m_pMsgReply->setHeaderToHost();
+        } else {
+            /* 2: message header is got, match message to cycle buffer */
+            m_pMsgReply->setBuf(recv_data.cur_addr);
+        }
```

Comment:
Logic could be simplified here. 

```suggestion
        } else {
            /* 2: message header is got, match message to cycle buffer */
            m_pMsgReply->setBuf(recv_data.cur_addr);
            if (recv_data.cur_offset < MsgHeader::EFFECTIVE_SIZE) {
                m_pMsgReply->setHeaderToHost();
            }
        }

```

Reply:
Same code, but more nested branches. I don't think that this is simpler.